bpf/dns_tracker.h

```diff
@@ -0,0 +1,141 @@
+/*
+    light weight DNS tracker using trace points.
+*/
+
+#ifndef __DNS_TRACKER_H__
+#define __DNS_TRACKER_H__
+#include "utils.h"
+
+#define DNS_PORT        53
+#define DNS_QR_FLAG     0x80
+
+struct dns_header {
+    u16 id;
+    u16 flags;
+    u16 qdcount;
+    u16 ancount;
+    u16 nscount;
+    u16 arcount;
+};
+
+struct dns_request_key {
+    u16 id;
+};
+
+struct dns_request_val {
+    u8 srcip[IP_MAX_LEN];
+};
+
+static inline int dns_addr_compare(u8 *s, u8 *d, u32 len) {
+    for (int i = 0; i < len; i++) {
+        if (s[i] != d[i]) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+static inline void find_or_create_dns_flow(flow_id *id, struct dns_header *dns, int len, int dir, u16 flags) {
+    id->direction = dir;
+    flow_metrics *aggregate_flow = bpf_map_lookup_elem(&aggregated_flows, id);
+    if (aggregate_flow != NULL) {
+        aggregate_flow->dns_record.id = bpf_ntohs(dns->id);
+        aggregate_flow->dns_record.flags = bpf_ntohs(dns->flags);
+        aggregate_flow->dns_record.req_mono_time_ts = bpf_ktime_get_ns();
+        bpf_map_update_elem(&aggregated_flows, &id, aggregate_flow, BPF_ANY);
+    } else {
+        // there is no matching flows so lets create new one and add the drops
+        u64 current_time = bpf_ktime_get_ns();
+         flow_metrics new_flow;
+        if (dir == INGRESS) {
+            new_flow.start_mono_time_ts = current_time;
+            new_flow.end_mono_time_ts = current_time;
+            new_flow.packets = 1;
+            new_flow.bytes = len;
+            new_flow.flags = flags;
+            new_flow.dns_record.id = bpf_ntohs(dns->id);
+            new_flow.dns_record.flags = bpf_ntohs(dns->flags);
+            new_flow.dns_record.req_mono_time_ts = current_time;
+        } else {
+            new_flow.start_mono_time_ts = current_time;
+            new_flow.end_mono_time_ts = current_time;
+            new_flow.packets = 1;
+            new_flow.bytes = len;
+            new_flow.flags = flags;
+            new_flow.dns_record.id = bpf_ntohs(dns->id);
+            new_flow.dns_record.flags = bpf_ntohs(dns->flags);
+            new_flow.dns_record.rsp_mono_time_ts = current_time;
+        }
+        bpf_map_update_elem(&aggregated_flows, id, &new_flow, BPF_ANY);
+    }
+}
+
+static inline int trace_dns(struct sk_buff *skb) {
+    struct sock *sk = skb->sk;
+    flow_id id;
+    u8 protocol = 0;
+    u16 family = 0,flags = 0, len = 0;
+
+    __builtin_memset(&id, 0, sizeof(id));
+
+    // pull in details from the packet headers and the sock struct
+    bpf_probe_read(&family, sizeof(u16), &sk->__sk_common.skc_family);
```

Comment:
Not all packets will have a socket associated with them, in which case this will fail. If you just want to extract the packet protocol type, you'll probably be better off looking at the `skb->protocol` field. You'll have to deal with VLAN tagging, though; see the `skb_protocol()` helper in the kernel source...

Reply:
fixed